node/network/approval-distribution/src/lib.rs

```diff
@@ -96,7 +105,145 @@ impl RecentlyOutdated {
+	approvals: HashMap<CandidateIndex, IndirectSignedApprovalVote>,
+	// The validator index of the assignment signer.
+	validator_index: ValidatorIndex,
+	// Information required for gossiping to other peers using the grid topology.
+	routing_info: ApprovalRouting,
+}
+
+#[derive(Debug)]
+enum ApprovalEntryError {
+	InvalidValidatorIndex,
+	CandidateIndexOutOfBounds,
+	InvalidCandidateIndex,
+	DuplicateApproval,
+}
+
+impl ApprovalEntry {
+	pub fn new(
+		assignment: IndirectAssignmentCertV2,
+		candidates: CandidateBitfield,
+		routing_info: ApprovalRouting,
+	) -> ApprovalEntry {
+		Self {
+			validator_index: assignment.validator,
+			assignment,
+			approvals: HashMap::with_capacity(candidates.len()),
+			candidates,
+			routing_info,
+		}
+	}
+
+	// Create a `MessageSubject` to reference the assignment.
+	pub fn create_assignment_knowledge(&self, block_hash: Hash) -> (MessageSubject, MessageKind) {
+		(
+			MessageSubject(block_hash, self.candidates.clone(), self.validator_index),
+			MessageKind::Assignment,
+		)
+	}
+
+	// Create a `MessageSubject` to reference the approval.
+	pub fn create_approval_knowledge(
+		&self,
+		block_hash: Hash,
+		candidate_index: CandidateIndex,
+	) -> (MessageSubject, MessageKind) {
+		(
+			MessageSubject(block_hash, candidate_index.into(), self.validator_index),
+			MessageKind::Approval,
+		)
+	}
+
+	// Updates routing information and returns the previous information if any.
+	pub fn routing_info_mut(&mut self) -> &mut ApprovalRouting {
+		&mut self.routing_info
+	}
+
+	// Get the routing information.
+	pub fn routing_info(&self) -> &ApprovalRouting {
+		&self.routing_info
+	}
+
+	// Update routing information.
+	pub fn update_required_routing(&mut self, required_routing: RequiredRouting) {
+		self.routing_info.required_routing = required_routing;
+	}
+
+	// Records a new approval. Returns false if the claimed candidate is not found or we already have received the approval.
+	pub fn note_approval(
+		&mut self,
+		approval: IndirectSignedApprovalVote,
+	) -> Result<(), ApprovalEntryError> {
+		// First do some sanity checks:
+		// - check validator index matches
+		// - check claimed candidate
+		// - check for duplicate approval
+		if self.validator_index != approval.validator {
+			return Err(ApprovalEntryError::InvalidValidatorIndex)
+		}
+
+		if self.candidates.len() <= approval.candidate_index as usize {
+			return Err(ApprovalEntryError::CandidateIndexOutOfBounds)
+		}
+
+		if !self.candidates.bit_at(approval.candidate_index.as_bit_index()) {
+			return Err(ApprovalEntryError::InvalidCandidateIndex)
+		}
+
+		if self.approvals.contains_key(&approval.candidate_index) {
+			return Err(ApprovalEntryError::DuplicateApproval)
+		}
+
+		self.approvals.insert(approval.candidate_index, approval);
+		Ok(())
+	}
+
+	// Get the assignment certiticate and claimed candidates.
+	pub fn get_assignment(&self) -> (IndirectAssignmentCertV2, CandidateBitfield) {
```

Comment:
naming nit (not only here): the `get_` part seems redundant, this doesn't follow the Rust API guidelines on naming: https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter